File: sdks/go/sigil/validation.go
package sigil

import (
	"errors"
	"fmt"
	"strings"
)

func ValidateGeneration(g Generation) error {
	if g.Mode != "" && g.Mode != GenerationModeSync && g.Mode != GenerationModeStream {
		return errors.New("generation.mode must be one of SYNC|STREAM")
	}

	if strings.TrimSpace(g.Model.Provider) == "" {
		return errors.New("generation.model.provider is required")
	}

	if strings.TrimSpace(g.Model.Name) == "" {
		return errors.New("generation.model.name is required")
	}

	for i := range g.Input {
		if err := validateMessage("generation.input", i, g.Input[i]); err != nil {
			return err
		}
	}

	for i := range g.Output {
		if err := validateMessage("generation.output", i, g.Output[i]); err != nil {
			return err
		}
	}

	for i := range g.Tools {
		if strings.TrimSpace(g.Tools[i].Name) == "" {
			return fmt.Errorf("generation.tools[%d].name is required", i)
		}
	}

	for i := range g.Artifacts {
		if err := validateArtifact(i, g.Artifacts[i]); err != nil {
			return err
		}
	}

	return nil
}

func validateMessage(path string, index int, message Message) error {
	switch message.Role {
	case RoleUser, RoleAssistant, RoleTool:
	default:
		return fmt.Errorf("%s[%d].role must be one of user|assistant|tool", path, index)
	}

	if len(message.Parts) == 0 {
		return fmt.Errorf("%s[%d].parts must not be empty", path, index)
	}

	for i := range message.Parts {
		if err := validatePart(path, index, i, message.Role, message.Parts[i]); err != nil {
			return err
		}
	}

	return nil
}

func validatePart(path string, messageIndex, partIndex int, role Role, part Part) error {
	switch part.Kind {
	case PartKindText, PartKindThinking, PartKindToolCall, PartKindToolResult:
	default:
		return fmt.Errorf("%s[%d].parts[%d].kind is invalid", path, messageIndex, partIndex)
	}

	fieldCount := 0
	if strings.TrimSpace(part.Text) != "" {
		fieldCount++
	}
	if strings.TrimSpace(part.Thinking) != "" {
		fieldCount++
	}
	if part.ToolCall != nil {
		fieldCount++
	}
	if part.ToolResult != nil {
		fieldCount++
	}

	if fieldCount != 1 {
		return fmt.Errorf("%s[%d].parts[%d] must set exactly one payload field", path, messageIndex, partIndex)
	}

	switch part.Kind {
	case PartKindText:
		if strings.TrimSpace(part.Text) == "" {
			return fmt.Errorf("%s[%d].parts[%d].text is required", path, messageIndex, partIndex)
		}
	case PartKindThinking:
		if role != RoleAssistant {
			return fmt.Errorf("%s[%d].parts[%d].thinking only allowed for assistant role", path, messageIndex, partIndex)
		}
		if strings.TrimSpace(part.Thinking) == "" {
			return fmt.Errorf("%s[%d].parts[%d].thinking is required", path, messageIndex, partIndex)
		}
	case PartKindToolCall:
		if role != RoleAssistant {
			return fmt.Errorf("%s[%d].parts[%d].tool_call only allowed for assistant role", path, messageIndex, partIndex)
		}
		if part.ToolCall == nil || strings.TrimSpace(part.ToolCall.Name) == "" {
			return fmt.Errorf("%s[%d].parts[%d].tool_call.name is required", path, messageIndex, partIndex)
		}
	case PartKindToolResult:
		if role != RoleTool {
			return fmt.Errorf("%s[%d].parts[%d].tool_result only allowed for tool role", path, messageIndex, partIndex)
		}
		if part.ToolResult == nil {
			return fmt.Errorf("%s[%d].parts[%d].tool_result is required", path, messageIndex, partIndex)
		}
	}

	return nil
}

func validateArtifact(index int, artifact Artifact) error {
	switch artifact.Kind {
	case ArtifactKindRequest, ArtifactKindResponse, ArtifactKindTools, ArtifactKindProviderEvent:
	default:
		return fmt.Errorf("generation.artifacts[%d].kind is invalid", index)
	}

	if strings.TrimSpace(artifact.RecordID) == "" && len(artifact.Payload) == 0 {
		return fmt.Errorf("generation.artifacts[%d] must provide payload or record_id", index)
	}

	return nil
}

File: sdks/python/sigil_sdk/validation.py
"""Generation payload validation logic."""

from __future__ import annotations

from .models import ArtifactKind, Generation, GenerationMode, MessageRole, PartKind


def validate_generation(generation: Generation) -> None:
    """Raises ValueError when a generation payload is invalid."""

    if generation.mode is not None and generation.mode not in (GenerationMode.SYNC, GenerationMode.STREAM):
        raise ValueError("generation.mode must be one of SYNC|STREAM")

    if generation.model.provider.strip() == "":
        raise ValueError("generation.model.provider is required")

    if generation.model.name.strip() == "":
        raise ValueError("generation.model.name is required")

    for index, message in enumerate(generation.input):
        _validate_message("generation.input", index, message.role.value if hasattr(message.role, "value") else str(message.role), message.parts)

    for index, message in enumerate(generation.output):
        _validate_message("generation.output", index, message.role.value if hasattr(message.role, "value") else str(message.role), message.parts)

    for index, tool in enumerate(generation.tools):
        if tool.name.strip() == "":
            raise ValueError(f"generation.tools[{index}].name is required")

    for index, artifact in enumerate(generation.artifacts):
        if artifact.kind not in (
            ArtifactKind.REQUEST,
            ArtifactKind.RESPONSE,
            ArtifactKind.TOOLS,
            ArtifactKind.PROVIDER_EVENT,
        ):
            raise ValueError(f"generation.artifacts[{index}].kind is invalid")
        if artifact.record_id.strip() == "" and len(artifact.payload) == 0:
            raise ValueError(f"generation.artifacts[{index}] must provide payload or record_id")


def _validate_message(path: str, index: int, role: str, parts: list[object]) -> None:
    if role not in (MessageRole.USER.value, MessageRole.ASSISTANT.value, MessageRole.TOOL.value):
        raise ValueError(f"{path}[{index}].role must be one of user|assistant|tool")

    if len(parts) == 0:
        raise ValueError(f"{path}[{index}].parts must not be empty")

    for part_index, part in enumerate(parts):
        _validate_part(path, index, part_index, role, part)


def _validate_part(path: str, message_index: int, part_index: int, role: str, part: object) -> None:
    kind = part.kind.value if hasattr(part.kind, "value") else str(part.kind)

    if kind not in (
        PartKind.TEXT.value,
        PartKind.THINKING.value,
        PartKind.TOOL_CALL.value,
        PartKind.TOOL_RESULT.value,
    ):
        raise ValueError(f"{path}[{message_index}].parts[{part_index}].kind is invalid")

    field_count = 0
    if getattr(part, "text", "").strip() != "":
        field_count += 1
    if getattr(part, "thinking", "").strip() != "":
        field_count += 1
    if getattr(part, "tool_call", None) is not None:
        field_count += 1
    if getattr(part, "tool_result", None) is not None:
        field_count += 1

    if field_count != 1:
        raise ValueError(f"{path}[{message_index}].parts[{part_index}] must set exactly one payload field")

    if kind == PartKind.TEXT.value:
        if getattr(part, "text", "").strip() == "":
            raise ValueError(f"{path}[{message_index}].parts[{part_index}].text is required")
        return

    if kind == PartKind.THINKING.value:
        if role != MessageRole.ASSISTANT.value:
            raise ValueError(f"{path}[{message_index}].parts[{part_index}].thinking only allowed for assistant role")
        if getattr(part, "thinking", "").strip() == "":
            raise ValueError(f"{path}[{message_index}].parts[{part_index}].thinking is required")
        return

    if kind == PartKind.TOOL_CALL.value:
        if role != MessageRole.ASSISTANT.value:
            raise ValueError(f"{path}[{message_index}].parts[{part_index}].tool_call only allowed for assistant role")
        tool_call = getattr(part, "tool_call", None)
        if tool_call is None or getattr(tool_call, "name", "").strip() == "":
            raise ValueError(f"{path}[{message_index}].parts[{part_index}].tool_call.name is required")
        return

    if role != MessageRole.TOOL.value:
        raise ValueError(f"{path}[{message_index}].parts[{part_index}].tool_result only allowed for tool role")
    if getattr(part, "tool_result", None) is None:
        raise ValueError(f"{path}[{message_index}].parts[{part_index}].tool_result is required")
